### PR TITLE
feat[apis]: Pass frontend to Snuba - phase 1

### PR DESCRIPTION
**Pull request by sentaur-athena**
EDIT: updated boolean is_frontend with query_source based on feedback
---

For better debugability we're passing a new field to snuba while querying: `is_frontend`. I decided the field is a better fit for request header since that's how we pass the `referrer` in the Snuba request and `is_frontend` is kind of a break down of referrer. 

This PR right now only apply the changes to one endpoint, putting it up for discussion before adding it everywhere else.

### CODE REVIEW

File: src/sentry/utils/snuba.py [hunk 1]
```diff
@@ -923,12 +933,17 @@ def get_cache_key(query: SnubaQuery) -> str:
...
+        bool | None
+    ) = None,  # TODO: @athena Make this field required after updated all the callsites
     use_cache: bool | None = False,
 ) -> ResultSet:
     headers = {}
     validate_referrer(referrer)
     if referrer:
         headers["referer"] = referrer
+    if is_frontend is not None:
+        headers["is_frontend"] = str(is_frontend)
```
Comment (evanh) [inline src/sentry/utils/snuba.py:946]:
```suggestion
        headers["is-frontend"] = str(is_frontend)
```

Small nit.

**Review by evanh — COMMENTED** [review]
I like this idea, I think it would be helpful.

Is the Sentry rate limiter already using this middleware?

**Comment by Zylphrex** [comment]
Instead of a boolean `is-frontend` header, what if it's called something like `query-source` and can take on values like
- `browser` for ones initiated by the frontend
- `api` for ones initiated by the api
- `job` for ones initiated by celery jobs in sentry
- etc

**Comment by sentaur-athena** [comment]
@evanh which middleware are you talking about? The API rate limiting is not directly related to this PR, this is milestone 1 of the [tech spec](https://www.notion.so/sentry/Ratelimitting-Clarity-Internal-and-External-229b6897639841b683a79e3a6e16a7ea) here for more details.

**Comment by sentaur-athena** [comment]
@Zylphrex shouldn't referrer be different for each of these? My thinking is that if the caller is different the referrer has to be different. I can only think of 1 place where the code is shared -> browser vs. backend API call so referrer will be the same and this flag is needed. I'm not familiar with all snuba calls though so open to feedback.

**Comment by Zylphrex** [comment]
> @Zylphrex shouldn't referrer be different for each of these? My thinking is that if the caller is different the referrer has to be different. I can only think of 1 place where the code is shared -> browser vs. backend API call so referrer will be the same and this flag is needed. I'm not familiar with all snuba calls though so open to feedback.

For most referrers, this is true but there are some cases where this assumption doesn't hold. For example, the `/api/0/organizations/{organization_id_or_slug}/events/` endpoint can be used by the frontend, directly as an API by a user, and I've seen some backend code that uses the endpoint as well. This endpoint uses a default referrer when there isn't a valid one specified so everything gets grouped into one. So it can be helpful to be able to distinguish between the various sources.

**Comment by sentaur-athena** [comment]
@Zylphrex, I think the issue you're talking about is that:
> This endpoint uses a default referrer when there isn't a valid one specified so everything gets grouped into one

I'm trying to go through all APIs that have snuba calls in them and fix the shared referrer as well. I'll look into the one you mentioned. Thanks for the context.

**Review by volokluev — COMMENTED** [review]
The `is_frontend` should go into the `tenant_ids` of the snuba Request. That's what we use to make capacity allocation decisions

**Comment by evanh** [comment]
> @evanh which middleware are you talking about? The API rate limiting is not directly related to this PR, this is milestone 1 of the [tech spec](https://www.notion.so/sentry/Ratelimitting-Clarity-Internal-and-External-229b6897639841b683a79e3a6e16a7ea) here for more details.

Is the Sentry API rate limiter also going to make use of this `is-frontend` field? So we can have different limits on non-frontend requests?

> I'm trying to go through all APIs that have snuba calls in them and fix the shared referrer as well. I'll look into the one you mentioned. Thanks for the context.

There are also shared modules that get called by many different endpoints. `[tsdb](https://github.com/getsentry/sentry/tree/master/src/sentry/tsdb)` is one example.

**Comment by sentaur-athena** [comment]
@volokluev I'll make that change. It's not needed in the header at all then?

**Comment by sentaur-athena** [comment]
@evanh 
> Is the Sentry API rate limiter also going to make use of this is-frontend field? So we can have different limits on non-frontend requests?

No, this is only a clarity/monitoring attempt so in case of incident we know what's happening. The is_frontend flag is not super accurate and we shouldn't use it to determine logic. Basically no matter how we define it a direct API call can mimic the behavior and pretend to be a frontend call.

File: src/sentry/snuba/query_sources.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from enum import Enum
+
+
+class QuerySource(Enum):
```
Comment (sentaur-athena) [inline src/sentry/snuba/query_sources.py:4]: will add more as I discover

**Review by volokluev — APPROVED** [review]

**Pull request merged**